File: go/pkg/bysquare/pay/validations.go
package pay

import (
	"fmt"

	"github.com/xseman/bysquare/go/pkg/bysquare"
)

// ValidationError represents a validation error with path information.
type ValidationError struct {
	Message string
	Path    string
}

func (e *ValidationError) Error() string {
	return fmt.Sprintf("%s (path: %s)", e.Message, e.Path)
}

// ValidateDataModel validates the complete data model.
func ValidateDataModel(model *DataModel, version ...bysquare.Version) error {
	v := bysquare.Version120
	if len(version) > 0 {
		v = version[0]
	}

	if len(model.Payments) == 0 {
		return &ValidationError{
			Message: "at least one payment required",
			Path:    "payments",
		}
	}

	for i, payment := range model.Payments {
		path := fmt.Sprintf("payments[%d]", i)
		if err := ValidateSimplePayment(&payment, path, v); err != nil {
			return err
		}
	}

	return nil
}

// ValidateSimplePayment validates a single payment.
func ValidateSimplePayment(payment *SimplePayment, path string, version bysquare.Version) error {
	if len(payment.BankAccounts) == 0 {
		return &ValidationError{
			Message: "at least one bank account required",
			Path:    path + ".bankAccounts",
		}
	}

	for i, account := range payment.BankAccounts {
		accountPath := fmt.Sprintf("%s.bankAccounts[%d]", path, i)
		if err := ValidateBankAccount(&account, accountPath); err != nil {
			return err
		}
	}

	if payment.CurrencyCode != "" {
		if !bysquare.IsValidCurrencyCode(string(payment.CurrencyCode)) {
			return &ValidationError{
				Message: "invalid currency code (ISO 4217)",
				Path:    path + ".currencyCode",
			}
		}
	}

	if payment.PaymentDueDate != "" {
		if !bysquare.IsValidDate(payment.PaymentDueDate) {
			return &ValidationError{
				Message: "invalid date format (YYYYMMDD per v1.2 specification)",
				Path:    path + ".paymentDueDate",
			}
		}
	}

	if payment.Type == PaymentTypeStandingOrder && payment.StandingOrderExt != nil {
		if payment.StandingOrderExt.LastDate != "" && !bysquare.IsValidDate(payment.StandingOrderExt.LastDate) {
			return &ValidationError{
				Message: "invalid date format (YYYYMMDD per v1.2 specification)",
				Path:    path + ".standingOrderExt.lastDate",
			}
		}
	}

	if payment.Type == PaymentTypeDirectDebit && payment.DirectDebitExt != nil {
		if payment.DirectDebitExt.ValidTillDate != "" && !bysquare.IsValidDate(payment.DirectDebitExt.ValidTillDate) {
			return &ValidationError{
				Message: "invalid date format (YYYYMMDD per v1.2 specification)",
				Path:    path + ".directDebitExt.validTillDate",
			}
		}
	}

	if version >= bysquare.Version120 && (payment.Beneficiary == nil || payment.Beneficiary.Name == "") {
		return &ValidationError{
			Message: "beneficiary name is required",
			Path:    path + ".beneficiary.name",
		}
	}

	return nil
}

// ValidateBankAccount validates IBAN and BIC.
func ValidateBankAccount(account *BankAccount, path string) error {
	if !bysquare.IsValidIBAN(account.IBAN) {
		return &ValidationError{
			Message: "invalid IBAN (ISO 13616)",
			Path:    path + ".iban",
		}
	}

	if account.BIC != "" && !bysquare.IsValidBIC(account.BIC) {
		return &ValidationError{
			Message: "invalid BIC (ISO 9362)",
			Path:    path + ".bic",
		}
	}

	return nil
}

File: typescript/src/pay/validations.ts
import validator from "validator";

import { ValidationError } from "../errors.js";
import { isValidDate } from "../field.js";
import { Version } from "../types.js";
import {
	BankAccount,
	DataModel,
	type Payment,
	PaymentOptions,
} from "./types.js";

const ErrorMessages = {
	IBAN: "Invalid IBAN. Make sure ISO 13616 format is used.",
	BIC: "Invalid BIC. Make sure ISO 9362 format is used.",
	CurrencyCode: "Invalid currency code. Make sure ISO 4217 format is used.",
	Date: "Invalid date. Make sure YYYYMMDD format is used.",
	BeneficiaryName: "Beneficiary name is required.",
} as const;

/**
 * validates bankAccount fields:
 * - iban (ISO 13616)
 * - bic (ISO 9362)
 */
export function validateBankAccount(
	bankAccount: BankAccount,
	path: string,
): void {
	if (!validator.isIBAN(bankAccount.iban)) {
		throw new ValidationError(ErrorMessages.IBAN, `${path}.iban`);
	}

	if (bankAccount.bic && !validator.isBIC(bankAccount.bic)) {
		throw new ValidationError(ErrorMessages.BIC, `${path}.bic`);
	}
}

/**
 * validate simple payment fields:
 * - currencyCode (ISO 4217)
 * - paymentDueDate (YYYYMMDD format per v1.2 specification)
 * - bankAccounts
 *
 * @see validateBankAccount
 */
export function validateSimplePayment(
	simplePayment: Payment,
	path: string,
	version: Version = Version["1.2.0"],
): void {
	for (const [index, bankAccount] of simplePayment.bankAccounts.entries()) {
		validateBankAccount(bankAccount, `${path}.bankAccounts[${index}]`);
	}

	if (simplePayment.currencyCode && !validator.isISO4217(simplePayment.currencyCode)) {
		throw new ValidationError(
			ErrorMessages.CurrencyCode,
			`${path}.currencyCode`,
		);
	}

	if (
		simplePayment.paymentDueDate
		&& !isValidDate(simplePayment.paymentDueDate)
	) {
		throw new ValidationError(
			ErrorMessages.Date,
			`${path}.paymentDueDate`,
		);
	}

	if (
		simplePayment.type === PaymentOptions.StandingOrder
		&& simplePayment.lastDate
		&& !isValidDate(simplePayment.lastDate)
	) {
		throw new ValidationError(
			ErrorMessages.Date,
			`${path}.lastDate`,
		);
	}

	if (
		simplePayment.type === PaymentOptions.DirectDebit
		&& simplePayment.validTillDate
		&& !isValidDate(simplePayment.validTillDate)
	) {
		throw new ValidationError(
			ErrorMessages.Date,
			`${path}.validTillDate`,
		);
	}

	if (version >= Version["1.2.0"] && !simplePayment.beneficiary?.name) {
		throw new ValidationError(
			ErrorMessages.BeneficiaryName,
			`${path}.beneficiary.name`,
		);
	}
}

/**
 * Validate `payments` field of dataModel.
 *
 * @see validateSimplePayment
 * @see ValidationError
 */
export function validateDataModel(
	dataModel: DataModel,
	version: Version = Version["1.2.0"],
): DataModel {
	for (const [index, payment] of dataModel.payments.entries()) {
		validateSimplePayment(payment, `payments[${index}]`, version);
	}

	return dataModel;
}
